### PR TITLE
[BACK-67] Added fork and issue history

### PR DESCRIPTION
**Pull request by 2015builder**
### Description of Issue
Added fork, issue and star history
They are implemented as full history (e.g. get RepoStarRecords) 

### How I implemented
Reformatted the star history function to work for issues and forks

### CODE REVIEW

**Review by thomas-woehrle — CHANGES_REQUESTED** [review]
Looks good. 

however two things are wrong with it: 
1. the dates seem to be ordered wrong
2. the issue count is not correct. Do you know why that is ?
for exmple: 
Screenshot 2023-06-14 at 09 18 16

File: packages/graphql_server/src/test/test.ts
Comment (pczern) [inline packages/graphql_server/src/test/test.ts:1]: name the test file according to what you test

File: packages/graphql_server/src/starHistory/forkHistory.ts [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+ * @param {string} token - Github access token
+ * @returns Promise<number> : A promise that resolves to the total count of forks for the repository
+ */
+async function getRepoForksCount(repo: string, token?: string): Promise<number> {
+  const response: AxiosResponse<{ forks_count: number }> = await axios.get(
+    `https://api.github.com/repos/${repo}`,
+    {
+      headers: {
+        Accept: 'application/vnd.github+json',
+        Authorization: token ? `token ${token}` : ''
```
Comment (pczern) [inline packages/graphql_server/src/starHistory/forkHistory.ts:49]: does null work? can you use default prop on token?

File: packages/graphql_server/src/starHistory/forkHistory.ts [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+ * @returns {Promise<number>} A promise that resolves to the total number of pages.
+ * @throws {object} Throws an error object if the request fails or the repository has no forks.
+ */
+async function getPageCount(repo: string, token: string) {
+  const patchRes: AxiosResponse<ForksData[]> = await getRepoForks(repo, token)
+
+  const headerLink: string = (patchRes.headers['link'] as string) || ''
+
+  let pageCount = 1
+  const regResult = /next.*&page=(\d*).*last/.exec(headerLink)
```
Comment (pczern) [inline packages/graphql_server/src/starHistory/forkHistory.ts:112]: explain regex

File: packages/graphql_server/src/starHistory/forkHistory.ts [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+      }
+    })
+  }
+
+  const forkAmount = await getRepoForksCount(repo, token)
+  forkRecordsMap.set(utils.getDateString(Date.now()), forkAmount)
+
+  const forkRecords: ForkRecord[] = []
+
+  forkRecordsMap.forEach((v, k) => {
```
Comment (pczern) [inline packages/graphql_server/src/starHistory/forkHistory.ts:174]: map, concat?

File: packages/graphql_server/src/starHistory/issueHistory.ts
Comment (pczern) [inline packages/graphql_server/src/starHistory/issueHistory.ts:1]: why is this a copy of the other file, what about DRY
find a solution

**Review by pczern — CHANGES_REQUESTED** [review]

File: packages/graphql_server/src/starHistory/forkHistory.ts (hunk 2 above)
Comment (ArnoClaude) [inline packages/graphql_server/src/starHistory/forkHistory.ts:112]: An explanatory comment for RegEx explanations would be nice. Something like:
`// Regex to extract the 'next' and 'last' page numbers from GitHub's pagination Link header`

Also, it looks like you are only using regResult[1], so you might as well make it:
`const lastPageNumber = /next.*&page=(\d*).*last/.exec(headerLink)[1]` :)

**Review by ArnoClaude — CHANGES_REQUESTED** [review]
Great job, Max! :)

File: packages/graphql_server/src/starHistory/issueHistory.ts [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+export async function getRepoIssueRecords(
+  repo: string,
+  token: string,
+  maxRequestAmount: number,
+  startPage?: number
+) {
+  const pageCount = await getPageCount(repo, token)
+
+  const requestPages: number[] = []
+  if (startPage == undefined) {
```
Comment (DogukanGun) [inline packages/graphql_server/src/starHistory/issueHistory.ts:74]: You check undefined, but you don’t check null please check null or just use isNullorUndefined function. I am not sure it is written. If it is not, you can add this

**Review by DogukanGun — CHANGES_REQUESTED** [review]

File: packages/graphql_server/src/starHistory/forkHistory.ts (hunk 2 above)
Comment (2015builder) [reply packages/graphql_server/src/starHistory/forkHistory.ts:112]: @ArnoClaude Thanks for the already written out explanation. I will implement it in the next commit

File: packages/graphql_server/src/starHistory/forkHistory.ts [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+/** Creates the partial fork history for a specific timeframe
+ * @param {string} repo - Name of the GitHub repository in the format "owner/repository".
+ * @param {string} token - Github access token
+ * @param {TimeFrame} timeFrame - time frame that the history takes into account
+ * @param {number} maxRequestAmount - Maximum number of API requests to make to retrieve the fork records.
+ * This value can be exceeded by up to 6 times if there aren't many pages being looked at for the history
+ * This mostly happens for short timeframes / not a lot of forks
+ * @returns {ForkRecord[]} - An array of `ForkRecord` objects representing the fork records.
+ */
+export async function partialHistory(
```
Comment (thomas-woehrle) [inline packages/graphql_server/src/starHistory/forkHistory.ts:104]: rename to partialForkhistory

File: packages/graphql_server/src/starHistory/issueHistory.ts [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+/** Creates the partial issue history for a specific timeframe
+ * @param {string} repo - Name of the GitHub repository in the format "owner/repository".
+ * @param {string} token - Github access token
+ * @param {TimeFrame} timeFrame - time frame that the history takes into account
+ * @param {number} maxRequestAmount - Maximum number of API requests to make to retrieve the issue records.
+ * This value can be exceeded by up to 6 times if there aren't many pages being looked at for the history
+ * This mostly happens for short timeframes / not a lot of issues
+ * @returns {IssueRecord[]} - An array of `IssueRecord` objects representing the issue records.
+ */
+export async function partialHistory(
```
Comment (thomas-woehrle) [inline packages/graphql_server/src/starHistory/issueHistory.ts:104]: rename to partialIssueHistory

File: packages/graphql_server/src/starHistory/starHistory.ts [hunk 7]
```diff
@@ -143,3 +85,103 @@ async function getRepoStarsMap(
...
+/** Creates the partial star history for a specific timeframe
+ * @param {string} repo - Name of the GitHub repository in the format "owner/repository".
+ * @param {string} token - Github access token
+ * @param {TimeFrame} timeFrame - time frame that the history takes into account
+ * @param {number} maxRequestAmount - Maximum number of API requests to make to retrieve the star records.
+ * This value can be exceeded by up to 6 times if there aren't many pages being looked at for the history
+ * This mostly happens for short timeframes / not a lot of stars
+ * @returns {StarRecord[]} - An array of `StarRecord` objects representing the star records.
+ */
+export async function partialHistory(
```
Comment (thomas-woehrle) [inline packages/graphql_server/src/starHistory/starHistory.ts:98]: rename to partialStarHistory

**Review by thomas-woehrle — CHANGES_REQUESTED** [review]
**1. folder name:** 
 rename the `/starHistory` folder to `/githubHistory`

**2. partial History**
running starHistory and partialStarHistory for the day and week for AntonOsika/gpt-engineer:
Screenshot 2023-06-17 at 11 45 25

For specifying day it also delivered entries older than a day. For specifying week it did not finish and was running a long time, all the while using API calls.
Something similar happend with partialForkHistory. They just need a insane amount of API calls. 

So I appreciate the effort but we might not need the partialHistories under these circumstances. I will write you on slack. 

**3. zero values**
Forks history works great normally, but when running it for `jst-seminar-rostlab-tum/truffle-ai-backend` which has zero forks an exception was thrown

**4. issue history**
We discussed it before, but when running `jst-seminar-rostlab-tum/truffle-ai-backend` it returns a history of issues (it says 8 tor today). But we never had even a single issue... Again: do you know why that is? or what this number shows. If we don't know that the issue history is not of high value.

**Review by thomas-woehrle — CHANGES_REQUESTED** [review]
I tested it again. 
The bug for forkHistory for not forked repos is fixed. 

But all lot of the problems with the partial histories exist still. 
They are incredibly expensive in API calls and do not really work well unless under ideal circumstances. 
Screenshot 2023-06-18 at 00 05 10
Screenshot 2023-06-18 at 00 07 02

Since they do not provide huge value, I would suggest that you delete them completely. We don't need them. 

The issue history is not needed as well. It provides no value when it is not accurate. And you said yourself, that you don't know what the values actually represent. So unless you know in the meanwhile, delete that as well. 
because we can not offer a issue history which clearly does not depict issues.

Good job on the fork history though.

**Review by thomas-woehrle — APPROVED** [review]

**Review by pczern — CHANGES_REQUESTED** [review]
3 files are heavily identical and don't adhere to DRY principle

**Comment by 2015builder** [comment]
> 3 files are heavily identical and don't adhere to DRY principle

@pczern You are correct that the file structure of the 3 history files is very similar but because they all work quite different in small details I don't think that there is a much more efficient way to structure the code nor think I that this would lead to improvements in readability or maintainability. So personally I would believe that keeping the structure as it is is the best way to proceed with these methods.

**Review by pczern — APPROVED** [review]

**Review by pczern — APPROVED** [review]
we @kigadye @thomas-woehrle decided to approve as we are right before handover

**Review by DogukanGun — APPROVED** [review]

**Comment by thomas-woehrle** [comment]
I will squash merge this, for the unlikely case that something fails

**Pull request merged**